File: 51-n-queens/51-n-queens.java
class Solution {
    private void solve(int col,List<List<String>> ans,char board[][],int []leftRow,int []lowerDiagonal,int []upperDiagonal)
{
if(col==board.length)
{
ans.add(construct(board));
return;
}
for(int row=0;row<board.length;row++)
{
if(leftRow[row]==0&&lowerDiagonal[row+col]==0&&upperDiagonal[board.length-1+col-row]==0)
{
board[row][col]='Q';
leftRow[row]=1;
lowerDiagonal[row+col]=1;
upperDiagonal[board.length-1+col-row]=1;
solve(col+1,ans,board,leftRow,lowerDiagonal,upperDiagonal);
board[row][col]='.';
leftRow[row]=0;
lowerDiagonal[row+col]=0;
upperDiagonal[board.length-1+col-row]=0;
}
}
}
private List<String> construct(char board[][])
{
List<String> res=new LinkedList<>();
for(int i=0;i<board.length;i++)
{
String s=new String(board[i]);
res.add(s);
}
return res;
}
public List<List<String>> solveNQueens(int n) {
char board[][]=new char[n][n];
for(int i=0;i<n;i++)
{
for(int j=0;j<n;j++)
{
board[i][j]='.';}}
List<List<String>> ans=new ArrayList<>();
int leftRow[]=new int[n];
int []lowerDiagonal=new int[2*n-1];
int []upperDiagonal=new int[2*n-1];
solve(0,ans,board,leftRow,lowerDiagonal,upperDiagonal);
return ans;
}
}

File: 51-n-queens/51-n-queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        col=set()
        pos_dig=set()
        neg_dig=set()
        ans=[]
        board=[["."]*n for i in range(n)]
        
        def backtrack(r):
            if r==n:
                copy=["".join(row) for row in board]
                ans.append(copy)
            for c in range(n):
                if c in col or (r+c) in pos_dig or (r-c) in neg_dig:
                    continue
                col.add(c)
                pos_dig.add(r+c)
                neg_dig.add(r-c)
                board[r][c]="Q"
                
                backtrack(r+1)
                
                col.remove(c)
                pos_dig.remove(r+c)
                neg_dig.remove(r-c)
                board[r][c]="."
        backtrack(0)
        return ans
                
        
        '''col=set()
        posdig=set()#(r+c)
        negdig=set()#(r-c)
        res=[]
        board=[["."]*n for i in range(n)]
    #backtracking function
        def backtrack(r):
            if r==n:
                copy=["".join(row) for row in board]
                res.append(copy)
                return
            for c in range(n):
                if c in col or (r+c) in posdig or (r-c) in negdig:
                    continue
                
                col.add(c)
                posdig.add(r+c)
                negdig.add(r-c)
                board[r][c]="Q"
                
                backtrack(r+1)
                
                col.remove(c)
                posdig.remove(r+c)
                negdig.remove(r-c)
                board[r][c]="."
        backtrack(0)
        return res'''